src/Microsoft.Health.Core/Features/Security/Authorization/RoleBasedAuthorizationService.cs

```diff
@@ -0,0 +1,62 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Health.Core.Configs;
+using Microsoft.Health.Core.Features.Context;
+
+namespace Microsoft.Health.Core.Features.Security.Authorization
+{
+    /// <summary>
+    /// Service used for checking if given set of dataActions are present in a given request contexts principal.
+    /// </summary>
+    /// <typeparam name="TEnum">Type representing the dataActions for the service</typeparam>
+    /// <typeparam name="TContext">Type representing the IRequestContext implementation for the service</typeparam>
+    public class RoleBasedAuthorizationService<TEnum, TContext> : IAuthorizationService<TEnum>
+        where TEnum : Enum
+        where TContext : IRequestContext
+    {
+        private readonly GenericRequestContextAccessor<TContext> _genericRequestContextAccessor;
+        private readonly string _rolesClaimName;
+        private readonly Dictionary<string, Role<TEnum>> _roles;
+
+        public RoleBasedAuthorizationService(AuthorizationConfiguration<TEnum> authorizationConfiguration, GenericRequestContextAccessor<TContext> genericRequestContextAccessor)
+        {
+            EnsureArg.IsNotNull(authorizationConfiguration, nameof(authorizationConfiguration));
+            _genericRequestContextAccessor = EnsureArg.IsNotNull(genericRequestContextAccessor, nameof(genericRequestContextAccessor));
+
+            _rolesClaimName = authorizationConfiguration.RolesClaim;
+            _roles = authorizationConfiguration.Roles.ToDictionary(r => r.Name, StringComparer.OrdinalIgnoreCase);
+        }
+
+        public ValueTask<TEnum> CheckAccess(TEnum dataActions, CancellationToken cancellationToken)
+        {
+            ClaimsPrincipal principal = _genericRequestContextAccessor.RequestContext.Principal;
+
+            ulong permittedDataActions = 0;
+            ulong dataActionsUlong = Convert.ToUInt64(dataActions, NumberFormatInfo.InvariantInfo);
```

Comment:
Instead of this approach, you can easily create a delegate that is compiled at runtime. Something like this:
``` csharp
private static readonly Func<TEnum, ulong> ConvertToULong = CreateConvertToULongFunc();

private static Func<TEnum, ulong> CreateConvertToULongFunc()
{
    var parameterExpression = Expression.Parameter(typeof(TEnum));
    return Expression.Lambda<Func<TEnum, UInt64>>(Expression.Convert(parameterExpression, typeof(UInt64)), parameterExpression).Compile();
}
```

... and another one for converting in the other direction

It should be significatly faster:
|          Method |       Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |                           |---------------- |-----------:|----------:|----------:|-------:|------:|------:|----------:|                           |    ConvertClass | 37.8900 ns | 0.7981 ns | 1.6302 ns | 0.0092 |     - |     - |      48 B |                           | ConvertDelegate |  0.9626 ns | 0.0574 ns | 0.0661 ns |      - |     - |     - |         - |